src/optimizer/rule/RemoveNoopProjectRule.cpp

```diff
@@ -0,0 +1,108 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/RemoveNoopProjectRule.h"
+
+#include "optimizer/OptContext.h"
+#include "optimizer/OptGroup.h"
+#include "planner/PlanNode.h"
+#include "planner/Query.h"
+
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> RemoveNoopProjectRule::kInstance =
+    std::unique_ptr<RemoveNoopProjectRule>(new RemoveNoopProjectRule());
+
+RemoveNoopProjectRule::RemoveNoopProjectRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& RemoveNoopProjectRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kProject);
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> RemoveNoopProjectRule::transform(
+    OptContext* octx,
+    const MatchedResult& matched) const {
+    const auto* projGroupNode = matched.node;
+    const auto* oldProjNode = projGroupNode->node();
+    DCHECK_EQ(oldProjNode->kind(), PlanNode::Kind::kProject);
+
+    TransformResult result;
+    result.eraseAll = true;
+    const auto* projGroup = projGroupNode->group();
+    const auto depGroups = projGroupNode->dependencies();
+    DCHECK_EQ(depGroups.size(), 1);
+    const auto* depGroup = depGroups.front();
+    const auto& groupNodes = depGroup->groupNodes();
+    for (auto* groupNode : groupNodes) {
+        auto* newNode = groupNode->node()->clone();
+        const auto& newColNames = newNode->colNames();
+        const auto& oldColNames = oldProjNode->colNames();
+        auto colsNum = newColNames.size();
+        if (colsNum != oldColNames.size()) {
+            return TransformResult::noTransform();
+        }
+        for (size_t i = 0; i < colsNum; ++i) {
+            if (newColNames[i].compare(oldColNames[i])) {
+                return TransformResult::noTransform();
+            }
+        }
```

Comment:
Emmm, it seems not enough that we eliminate the project by analysing the column names. What if we had such a situation: (name, age) -> ($-.name as name, $-.name as age)? We should check the column expression too.

Reply:
It won't be matched when there are aliases in `YieldColumns`(line 95).